### DISCUSSION TITLE
New tab preserves environment variable from parent's startup, not current values

### DISCUSSION THREAD

**Discussion by jdholtz** [kovidgoyal/kitty#6581] [Q&A] [answered]
Hello. Currently, when I open a new tab in Kitty, the tab's shell preserves environment variables from when the first window (the terminal the tab was created from) started up, not its current environment variable values.

To explain the issue better, I detailed a demonstration below. This may just be a misunderstanding I have of Kitty tabs and/or Bash, but I would like to know if there is anything I can do to prevent this. Note that shell integration being enabled or disabled doesn't affect this behavior. 

Start a Kitty terminal with the `TEST` environment variable set:
```shell
TEST=1 kitty
```

In the new terminal created, unset the variable `TEST`:
```shell
unset TEST  # export TEST= leads to the same issue too
```

Create a new tab from the same terminal. In the new terminal, see how the `TEST` variable is still 1.
```shell
echo $TEST
1
```

Is there any way to prevent TEST from still being 1 in the new tab even though it was unset before creating the tab?

**Comment by kovidgoyal** [comment, marked as answer]
It doesnt preserve it from the first window, it preserves it for the
parent processes environment. If you want to clone the environment of a
particular window, run the clone-in-kitty command. But note that this
will not preserver unset as the final environment is the union of the
parent environment and the env in the window being cloned.

On Sat, Aug 26, 2023 at 08:17:15AM -0700, Joey Holtzman wrote:
> Hello. Currently, when I open a new tab in Kitty, the tab's shell preserves environment variables from when the first window (the terminal the tab was created from) started up, not its current environment variable values.
> 
> To explain the issue better, I detailed a demonstration below. This may just be a misunderstanding I have of Kitty tabs and/or Bash, but I would like to know if there is anything I can do to prevent this. Note that shell integration being enabled or disabled doesn't affect this behavior. 
> 
> Start a Kitty terminal with the `TEST` environment variable set:
> ```shell
> TEST=1 kitty
> ```
> 
> In the new terminal created, unset the variable `TEST`:
> ```shell
> unset TEST  # export TEST= leads to the same issue too
> ```
> 
> Create a new tab from the same terminal. In the new terminal, see how the `TEST` variable is still 1.
> ```shell
> echo $TEST
> 1
> ```
> 
> Is there any way to prevent TEST from still being 1 in the new tab even though it was unset before creating the tab? 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6581
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Reply by jdholtz** [reply]
That make sense that it preserves it from the parent process. I’ll use the clone-in-kitty command in this case. Thanks!
